# Life Science Computing

## Aufgabenstellung <a id="task"></a>

> Es wird eine umfassende Analyse eines Datensatzes erwartet. Dazu kann eine Progarmmiersprache eigener Wahl genutzt werden. Die Analyse wird als digitales Textdokument oder als ein <span style="color:blue">(Jupyter) Notebook </span>erstellt.

### Bewertungskriterien: <a id="evaluation-criteria"></a>
 * Korrektheit der Analyse
 * umfassende Datenuntersuchung
 * Beschreibung der Daten
 * Diskussion der Ergebnisse
 * kritische Bewertung der Ergebnisse
 * Einhaltung der Regeln guter wissenschaftlicher Praxis

## Datensatz <a id="dataset"></a>

**"Transcription profiling by high throughput sequencing in different developmental stages of Zea mays subsp. mays tissues"** <br>
Set: RNA-Seq mRNA baseline <br>
Organism: *Zea mays*<br>
Data Source: [Expression Atlas EBI Datensatz](https://www.ebi.ac.uk/gxa/experiments/E-MTAB-4342)

<img src="http://beta.floranorthamerica.org/w/images/a/ae/FNA25_P274_Zea_Mays_pg_702.jpeg" alt="Zea Mays" width="600"/>

***

# Inhaltsverzeichnis  <a id="toc"></a>

1. [Aufgabenstellung](#task)
    1. [Bewertungskriterien](#evaluation-criteria)
2. [Datensatz](#dataset)
3. [Inhaltsverzeichnis](#toc)
3. [Preparation](#prep)
    1. [Imports](#imports)
    2. [Load Data](#load-data)
4. [Exploring and Processing the Experimental Data](#exploring-exp)
    1. [Print all Columns](#print-cols)
    2. [Check if Columns are equal](#double-cols-exp)
    3. [Delete Data](#del-exp)
    4. [Exploring the Leftover-Data](#expl-leftover-exp)
    5. [<span style="color:yellow">Zwischenfazit zu den Experiment-Daten</span> 🏁](#fazit-leftover-exp)
5. [Exploring and Processing the Results](#exploring-results)
    1. [Print all Columns](#print-cols-results)
    2. [Check if Columns are equal](#double-cols-results)
    3. [Delete Data](#del-results)
    4. [Exploring the Leftover-Data](#expl-leftover-results)
    5. [<span style="color:yellow">Zwischenfazit zu den Ergebnis-Daten</span> 🏁](#fazit-leftover-results)




***

# Preparation <a id="prep"></a>

## Imports <a id="imports"></a>

In [1]:
import matplotlib.pyplot as plt
import matplotlib as mpl
import numpy as np
import pandas as pd

### Load Data <a id="load-data"></a>

In [2]:
exp_design = pd.read_csv('E-MTAB-4342-experiment-design.tsv', sep='\t', header=0)
results = pd.read_csv('E-MTAB-4342-query-results.fpkms.tsv', sep='\t', header=4)

***

# Exploring and Processing the Experimental Data <a id="exploring-exp"></a>

## Print all Columns <a id="print-cols"></a>

In [3]:
exp_design.columns

Index(['Run', 'Sample Characteristic[cultivar]',
       'Sample Characteristic Ontology Term[cultivar]',
       'Sample Characteristic[developmental stage]',
       'Sample Characteristic Ontology Term[developmental stage]',
       'Sample Characteristic[organism]',
       'Sample Characteristic Ontology Term[organism]',
       'Sample Characteristic[organism part]',
       'Sample Characteristic Ontology Term[organism part]',
       'Factor Value[developmental stage]',
       'Factor Value Ontology Term[developmental stage]',
       'Factor Value[organism part]',
       'Factor Value Ontology Term[organism part]', 'Analysed'],
      dtype='object')

## Check if Columns are equal <a id="double-cols-exp"></a>

In [4]:
exp_design['Sample Characteristic Ontology Term[developmental stage]'].equals(exp_design['Factor Value Ontology Term[developmental stage]'])

True

In [5]:
exp_design['Sample Characteristic Ontology Term[organism part]'].equals(exp_design['Factor Value Ontology Term[organism part]'])

True

In [6]:
exp_design['Sample Characteristic[organism part]'].equals(exp_design['Factor Value[organism part]'])

True

In [7]:
exp_design['Sample Characteristic[developmental stage]'].equals(exp_design['Factor Value[developmental stage]'])

True

## Delete Data <a id="del-exp"></a>

### Delete all Columns with less than 2 different Values <a id="del-ununique-exp"></a>

In [8]:
for col in exp_design.columns:
    if len(exp_design[col].unique()) == 1:
        del exp_design[col]

In [9]:
exp_design.columns

Index(['Run', 'Sample Characteristic[developmental stage]',
       'Sample Characteristic Ontology Term[developmental stage]',
       'Sample Characteristic[organism part]',
       'Sample Characteristic Ontology Term[organism part]',
       'Factor Value[developmental stage]',
       'Factor Value Ontology Term[developmental stage]',
       'Factor Value[organism part]',
       'Factor Value Ontology Term[organism part]', 'Analysed'],
      dtype='object')

### Delete double Columns <a id="del-dc-exp"></a>

In [10]:
del exp_design['Factor Value Ontology Term[developmental stage]']
del exp_design['Factor Value Ontology Term[organism part]']
del exp_design['Factor Value[organism part]']
del exp_design['Factor Value[developmental stage]']

### Delete Sample Characteristic Ontology Columns <a id="del-sco-exp"></a>

In [11]:
del exp_design['Sample Characteristic Ontology Term[developmental stage]']
del exp_design['Sample Characteristic Ontology Term[organism part]']

### Delete useless Cols <a id="del-ul-exp"></a>

In [12]:
del exp_design['Analysed']
del exp_design['Run']

In [13]:
exp_design = exp_design.rename(columns={"Sample Characteristic[developmental stage]": "developmental stage", "Sample Characteristic[organism part]": "organism part"})
exp_design

developmental stage organism part
0     6 days after pollination          leaf
1     6 days after pollination          leaf
2        6 days after anthesis          leaf
3    12 days after pollination          leaf
4    12 days after pollination          leaf
..                         ...           ...
265  16 days after pollination     endosperm
266  16 days after pollination  plant embryo
267  16 days after pollination  plant embryo
268  16 days after pollination  plant embryo
269  18 days after pollination          seed

[270 rows x 2 columns]

## Exploring the Leftover-Data <a id="expl-leftover-exp"></a>

In [14]:
print(exp_design["organism part"].nunique(), exp_design["developmental stage"].nunique())

44 32


In [15]:
exp_design.describe()

developmental stage organism part
count                   267           270
unique                   32            44
top     7 days after sowing          seed
freq                     24            33

## <span style="color:yellow">Zwischenfazit zu den Experiment-Daten 🏁</span><a id="fazit-leftover-exp"></a>

### Kultursorte
Um die Ergebnisse nicht aufgrund verschiedenen Genbestands zu verfälschen, wurde eine Inzuchtline der Sorte B73 gezüchtet. Die Inzuchtline stellt sicher, dass der Genbestand aller Probenobjekte gleich ist.
### Probenort / Organismus Teil
Die Expressionsrate sollte außerdem in verschiedenen Teilen des Organismus gemessen werden.</br>
Es wurden 44 verschiedene Abschnitte gemessen. Unter Anderem:
- Blatt
- Internodium
- verschiedene Pfahlwurzelzonen
- Stele
- Hauptwurzel
- Perikarp
- Seminalwurzel

### Entwicklungsstadien 🌱
Das eigentliche Ziel des Versuches war es, die Genexpressionsrate in Abhängigkeit von dem Entwicklungsstadium der Pflanze zu untersuchen. Es wurden 32 verschiedene Entwicklungsstadien untersucht.</br>
Unter Anderem verschiedene Zeitabstände nach:
- Bestäubung
- Anthese
- Aussaat

Außerdem wurde auch die Anzahl der sichbaren Blätter als Stadium eingeteilt und danach unterschieden.</br>
Zusätzlich wurden berücksichtigt:</br>
- Blütephase der ganzen Pflanze
- Stadium der Fruchtbildung der ganzen Pflanze in verschiedenen prozentualen Abschnitten

***

# Exploring and Processing the Results FPKM (Fragments Per Kilobase Million) <a id="exploring-results"></a>


## Print all Columns <a id="print-cols-results"></a>

In [16]:
results

Gene ID        Gene Name  root, 3 days after sowing  \
0      Zm00001eb000010  Zm00001eb000010                        2.0   
1      Zm00001eb000020  Zm00001eb000020                       24.0   
2      Zm00001eb000030  Zm00001eb000030                        NaN   
3      Zm00001eb000040  Zm00001eb000040                        NaN   
4      Zm00001eb000050  Zm00001eb000050                        NaN   
...                ...              ...                        ...   
37839  Zm00001eb442870  Zm00001eb442870                        NaN   
37840  Zm00001eb442890  Zm00001eb442890                        NaN   
37841  Zm00001eb442910  Zm00001eb442910                        NaN   
37842  Zm00001eb442960  Zm00001eb442960                        NaN   
37843  Zm00001eb443030  Zm00001eb443030                        NaN   

       differentiation zone of primary root, 3 days after sowing  \
0                                                    2.0           
1                                                   15.0           
2                                                    NaN           
3                                                    NaN           
4                                                    NaN           
...                                                  ...           
37839                                                NaN           
37840                                                NaN           
37841                                                NaN           
37842                                                NaN           
37843                                                NaN           

       meristematic zone and elongation zone, 3 days after sowing  \
0                                                    2.0            
1                                                   38.0            
2                                                    NaN            
3                                                    NaN            
4                                                    NaN            
...                                                  ...            
37839                                                NaN            
37840                                                NaN            
37841                                                NaN            
37842                                                0.2            
37843                                                NaN            

       stele, 3 days after sowing  \
0                             2.0   
1                            19.0   
2                             NaN   
3                             0.5   
4                             NaN   
...                           ...   
37839                         NaN   
37840                         NaN   
37841                         NaN   
37842                         NaN   
37843                         NaN   

       cortical parenchyma of root, 3 days after sowing  \
0                                                   3.0   
1                                                  11.0   
2                                                   NaN   
3                                                   NaN   
4                                                   NaN   
...                                                 ...   
37839                                               NaN   
37840                                               NaN   
37841                                               NaN   
37842                                               0.2   
37843                                               NaN   

       coleoptile, 6 days after sowing  primary root, 6 days after sowing  \
0                                  4.0                                3.0   
1                                 38.0                               31.0   
2                                  0.1                                0.6   
3                                  NaN                            

## Check if Columns are equal <a id="double-cols-results"></a>


In [17]:
results['Gene ID'].equals(results['Gene Name'])

True

## Delete Data <a id="del-results"></a>

### Delete double Columns <a id="del-dc-results"></a>

In [18]:
del results['Gene Name']

In [19]:
results

Gene ID  root, 3 days after sowing  \
0      Zm00001eb000010                        2.0   
1      Zm00001eb000020                       24.0   
2      Zm00001eb000030                        NaN   
3      Zm00001eb000040                        NaN   
4      Zm00001eb000050                        NaN   
...                ...                        ...   
37839  Zm00001eb442870                        NaN   
37840  Zm00001eb442890                        NaN   
37841  Zm00001eb442910                        NaN   
37842  Zm00001eb442960                        NaN   
37843  Zm00001eb443030                        NaN   

       differentiation zone of primary root, 3 days after sowing  \
0                                                    2.0           
1                                                   15.0           
2                                                    NaN           
3                                                    NaN           
4                                                    NaN           
...                                                  ...           
37839                                                NaN           
37840                                                NaN           
37841                                                NaN           
37842                                                NaN           
37843                                                NaN           

       meristematic zone and elongation zone, 3 days after sowing  \
0                                                    2.0            
1                                                   38.0            
2                                                    NaN            
3                                                    NaN            
4                                                    NaN            
...                                                  ...            
37839                                                NaN            
37840                                                NaN            
37841                                                NaN            
37842                                                0.2            
37843                                                NaN            

       stele, 3 days after sowing  \
0                             2.0   
1                            19.0   
2                             NaN   
3                             0.5   
4                             NaN   
...                           ...   
37839                         NaN   
37840                         NaN   
37841                         NaN   
37842                         NaN   
37843                         NaN   

       cortical parenchyma of root, 3 days after sowing  \
0                                                   3.0   
1                                                  11.0   
2                                                   NaN   
3                                                   NaN   
4                                                   NaN   
...                                                 ...   
37839                                               NaN   
37840                                               NaN   
37841                                               NaN   
37842                                               0.2   
37843                                               NaN   

       coleoptile, 6 days after sowing  primary root, 6 days after sowing  \
0                                  4.0                                3.0   
1                                 38.0                               31.0   
2                                  0.1                                0.6   
3                                  NaN                                0.5   
4                                  NaN                                NaN   
...                                ...                                ...   
37839                              NaN 

## Exploring the Leftover-Data <a id="expl-leftover-results"></a>

### Data: **After Sowing**

In [20]:
after_sowing = results.filter(
    regex='days after sowing')
after_sowing

root, 3 days after sowing  \
0                            2.0   
1                           24.0   
2                            NaN   
3                            NaN   
4                            NaN   
...                          ...   
37839                        NaN   
37840                        NaN   
37841                        NaN   
37842                        NaN   
37843                        NaN   

       differentiation zone of primary root, 3 days after sowing  \
0                                                    2.0           
1                                                   15.0           
2                                                    NaN           
3                                                    NaN           
4                                                    NaN           
...                                                  ...           
37839                                                NaN           
37840                                                NaN           
37841                                                NaN           
37842                                                NaN           
37843                                                NaN           

       meristematic zone and elongation zone, 3 days after sowing  \
0                                                    2.0            
1                                                   38.0            
2                                                    NaN            
3                                                    NaN            
4                                                    NaN            
...                                                  ...            
37839                                                NaN            
37840                                                NaN            
37841                                                NaN            
37842                                                0.2            
37843                                                NaN            

       stele, 3 days after sowing  \
0                             2.0   
1                            19.0   
2                             NaN   
3                             0.5   
4                             NaN   
...                           ...   
37839                         NaN   
37840                         NaN   
37841                         NaN   
37842                         NaN   
37843                         NaN   

       cortical parenchyma of root, 3 days after sowing  \
0                                                   3.0   
1                                                  11.0   
2                                                   NaN   
3                                                   NaN   
4                                                   NaN   
...                                                 ...   
37839                                               NaN   
37840                                               NaN   
37841                                               NaN   
37842                                               0.2   
37843                                               NaN   

       coleoptile, 6 days after sowing  primary root, 6 days after sowing  \
0                                  4.0                                3.0   
1                                 38.0                               31.0   
2                                  0.1                                0.6   
3                                  NaN                                0.5   
4                                  NaN                                NaN   
...                                ...                                ...   
37839                              NaN                                NaN   
37840                              0.2                                NaN   
37841                              NaN                                NaN   
3784

#### Calculate how many Genes where not expressed after sowing compared to Total

In [21]:
len_all = len(results.index)

after_sowing = after_sowing.dropna(axis=0, how='all')
len_sowing = len(after_sowing.index)
print(len_all- len_sowing)

7213


#### Count how many Genes where expressed after sowing by days after sowing and Organism Part

In [22]:
after_sowing.count()

root, 3 days after sowing                                     24582
differentiation zone of primary root, 3 days after sowing     24787
meristematic zone and elongation zone, 3 days after sowing    22893
stele, 3 days after sowing                                    23762
cortical parenchyma of root, 3 days after sowing              25025
coleoptile, 6 days after sowing                               25759
primary root, 6 days after sowing                             25631
primary root, 7 days after sowing                             26315
root, 7 days after sowing                                     26103
seminal root, 7 days after sowing                             25064
taproot zone 1, 7 days after sowing                           23977
taproot zone 2, 7 days after sowing                           25058
taproot zone 3, 7 days after sowing                           25998
taproot zone 4, 7 days after sowing                           26809
dtype: int64

---

### Data: **After Pollination**

In [23]:
after_pollination = results.filter(
    regex='days after pollination')
after_pollination

seed, 2 days after pollination  seed, 4 days after pollination  \
0                                 5.0                             6.0   
1                                42.0                            40.0   
2                                 NaN                             0.2   
3                                 0.2                             0.1   
4                                 0.1                             NaN   
...                               ...                             ...   
37839                             NaN                             NaN   
37840                             0.2                             NaN   
37841                             NaN                             NaN   
37842                             2.0                             3.0   
37843                             NaN                             NaN   

       leaf, 6 days after pollination  internode, 6 days after pollination  \
0                                14.0                                 11.0   
1                                 0.9                                  2.0   
2                                 0.4                                  NaN   
3                                 NaN                                  NaN   
4                                 NaN                                  NaN   
...                               ...                                  ...   
37839                             NaN                                  NaN   
37840                             NaN                                  NaN   
37841                             NaN                                  NaN   
37842                             0.4                                  0.3   
37843                             NaN                                  NaN   

       seed, 6 days after pollination  seed, 8 days after pollination  \
0                                 5.0                             4.0   
1                                30.0                            29.0   
2                                 0.2                             NaN   
3                                 0.1                             0.1   
4                                 0.1                             NaN   
...                               ...                             ...   
37839                             NaN                             NaN   
37840                             NaN                             NaN   
37841                             NaN                             NaN   
37842                             3.0                             2.0   
37843                             NaN                             NaN   

       seed, 10 days after pollination  leaf, 12 days after pollination  \
0                                  4.0                             12.0   
1                                 28.0                              0.6   
2                                  0.1                              NaN   
3                                  0.1                              0.1   
4                                  NaN                              NaN   
...                                ...                              ...   
37839                              NaN                              NaN   
37840                              0.2                              NaN   
37841                              NaN                              NaN   
37842                              2.0                              0.2   
37843                              NaN                              NaN   

       internode, 12 days after pollination  seed, 12 days after pollination  \
0                                       6.0                              3.0   
1                                       4.0                             41.0   
2                                       NaN                              NaN   
3                                       NaN                              NaN   
4              

#### Calculate how many Genes where not expressed after pollination compared to Total

In [24]:
len_all = len(results.index)

after_pollination = after_pollination.dropna(axis=0, how='all')
len_pollination = len(after_pollination.index)
print(len_all- len_pollination)

2077


#### Calculate how many more Genes where expressed compared to sowing

In [25]:
print(len_pollination - len_sowing)

5136


#### Count how many Genes where expressed after pollination by days after pollination and Organism Part

In [26]:
after_pollination.count()

seed, 2 days after pollination             26907
seed, 4 days after pollination             27090
leaf, 6 days after pollination             26035
internode, 6 days after pollination        25390
seed, 6 days after pollination             26935
seed, 8 days after pollination             27191
seed, 10 days after pollination            27919
leaf, 12 days after pollination            25413
internode, 12 days after pollination       26289
seed, 12 days after pollination            25672
endosperm, 12 days after pollination       24969
seed, 14 days after pollination            26669
endosperm, 14 days after pollination       23174
seed, 16 days after pollination            25538
plant embryo, 16 days after pollination    24076
endosperm, 16 days after pollination       22740
leaf, 18 days after pollination            25620
internode, 18 days after pollination       25385
seed, 18 days after pollination            26032
plant embryo, 18 days after pollination    25016
endosperm, 18 days a

#### Compare Genes from Leaf by Days

In [27]:
after_pollination.filter(
    regex='leaf').count()


leaf, 6 days after pollination     26035
leaf, 12 days after pollination    25413
leaf, 18 days after pollination    25620
leaf, 24 days after pollination    25702
leaf, 30 days after pollination    26485
dtype: int64

#### Compare Genes from Endosperm by Days

In [28]:
after_pollination.filter(
    regex='endosperm').count()


endosperm, 12 days after pollination    24969
endosperm, 14 days after pollination    23174
endosperm, 16 days after pollination    22740
endosperm, 18 days after pollination    24444
endosperm, 20 days after pollination    23508
endosperm, 22 days after pollination    23009
endosperm, 24 days after pollination    22742
dtype: int64

## <span style="color:yellow">Zwischenfazit zu den Ergebnis-Daten 🏁</span><a id="fazit-leftover-results"></a>

### Nach der Aussaat
Die Messungen der Genexpression wurden an mehreren Tagen durchgeführt.
Jeweils 3, 6 und 7 Tage nach der Aussaat.

Die Probenentnahme 3 Tage nach der Aussaat erfolgte an folgenden Organismus-Teilen:
- Wurzel
- Differenzierungszone der Primärwurzel
- meristematische Zone und Streckungszone
- Stele
- Rindenparenchym der Wurzel

Am 6. Tag nach der Aussat wurden folgende Teile untersucht:
- Koleoptile
- Primärwurzel

Einen weiteren Tag später fanden die letzten Messungen dieses Entwicklungsstadiums in folgenden Teilen statt:
- Primärwurzel
- Wurzel
- Seminalwurzel
- Pfahlwurzel Zonen 1 - 4

Nach der Aussaat wurden 7213 Gene weniger expremiert, verglichen mit allen Wachstumsstadien in Summe.<br>
Mit Anzahl der Tage nach Aussaat stieg außerdem die Anzahl der expr. Gene.<br>
Nach 3 Tagen lag die Anzahl bei 22893 - 25025, bei 6 Tagen bei 25631-25759 und bei 7 Tagen bei 23977-26809.<br>
Es ist ein deutlicher Anstieg erkennbar. Unklar ist, ob dieser tatsächlich durch die Anzahl vergangener Tage und somit durch die Wachstumsphase bedingt ist, oder auf die unterschiedlichen Probenentnahmeorte zurückzuführen ist.<br>
Zwei Proben wurden an verschiedenen Tagen in der selben Zone entnommen. Diese legen die Vermutung nahe, dass die höhere Anzahl der Tage nach Aussaat zu einer höheren Expremierung führt.
Bei den Proben handelt es sich um die folgenden:

| **Tag** | **Primärwurzel** | **Wurzel** |
|:-------:|:----------------:|:----------:|
|    3    |                  |    24582   |
|    6    |       25631      |            |
|    7    |       26315      |    26103   |


### Nach der Bestäubung
Die Messungen wurden nach der Bestäubung jeweils im Abstand von 2 Tagen, bis Tag 30 durchgeführt.
Mit fortschreitendem Wachstumsstadium wurden die Messungen auch an verschiedenen Orten durchgeführt.
Getestet wurden z.B.:
- Samen
- Blatt
- Internodium
- Endosperm
- Pflanzenembryo
- Perikarp

Verglichen mit der Aussaat wurden 5136 Gene mehr exprimiert.
Die Anzahl der exprimierten Gene scheint hier keinen direkten Zusammenhang mit der Anzahl vergangener Tage im Entwicklungsstadium zu haben. Dies lässt sich aber nicht sicher sagen, da an verschiedenen Orten die Proben entnommen wurden.<br>
Die Proben der Blätter wurden im regelmäßigen Abstand von 6 Tagen genommen. Die Daten zeigen hier, dass es keinen linearen Zusammenhang gibt.

| **Tage nach der Bestäubung** | **Anzahl exprimierter Gene** |
|:----------------------------:|:----------------------------:|
|               6              |             26035            |
|              12              |             25413            |
|              18              |             25620            |
|              24              |             25702            |
|              30              |             26485            |


Um sicher zu gehen, dass dies nicht nur eine Eigenschaft der Blätter ist, habe ich zusätzlich die Daten des Endosperms tageweise gefiltert. Auch hier zeigt sich, dass es keinen linearen Zusammenhang gibt.

| **Tage nach der Bestäubung** | **Anzahl exprimierter Gene** |
|:----------------------------:|:----------------------------:|
|              12              |             24969            |
|              14              |             23174            |
|              16              |             22740            |
|              18              |             24444            |
|              20              |             23508            |
| 22                           |             23009            |
| 24                           |             22742            |


In [29]:
import numpy as np
from pandas import DataFrame
import seaborn as sns
# %matplotlib inline
results = results.fillna(0)
results.set_index('Gene ID', inplace=True)


